### DISCUSSION TITLE
Fixing a ShaderToy shader (lens distortion).

### DISCUSSION THREAD

**Discussion by MarioMey** [Limeth/obs-shaderfilter-plus#22] [Q&A]
I like this shader: https://www.shadertoy.com/view/3tXczj#. It creates like a distorted and misterious magic ball. Like this:
![imagen](https://user-images.githubusercontent.com/4954109/121445765-7d3f1500-c968-11eb-8dce-3ae289c1c9bb.png)
I converted to use it in obs-shaderfilter-plus. I get to this:

```glsl
#define DARK_EDGES

float4 render(vec2 uv )
{
   	vec2 disorsion = uv - 0.5;
	float aspect = builtin_uv_size.x/builtin_uv_size.y;
	
	disorsion.x *= aspect; // aspect correction
	
	// take distance from center
   	float len = length(disorsion);
	
	// these are the lens parameters
	float k1 = 1.2;
	float k2 = 1.0;
	float k3 = -3.2;
	
	disorsion 
		= disorsion*k1
		+ disorsion*len*k2 
		+ disorsion*len*len*k3;
		// higher powers may be added if necessary
   
	disorsion.x /= aspect; // aspect correction
	
	vec4 col = texture(image, disorsion + 0.5);
	
	#ifdef DARK_EDGES
	{
		float edge = 0.7;
		float dispersion = 0.03;
		col *= vec4(
			pow(max(edge-len, 0.0), 0.2),
			pow(max(edge-dispersion-len, 0.0), 0.2),
			pow(max(edge-dispersion*2.0-len, 0.0), 0.2),
		1)*1.2;
	}
	#endif

	return col;
}
```
But, instead of the same effect, I get this 🤦:
![imagen](https://user-images.githubusercontent.com/4954109/121445820-9b0c7a00-c968-11eb-9730-14590f47ba3a.png)

I think the main problem is about aspect. But I tried different things and I didn't arrive to anywhere.

Where is the error?

**Comment by Limeth** [comment]
Strange! Looks like "builtin_uv_size" has a bug!
It works fine for me when I set `aspect = 16.0/9.0;`, which is the aspect ratio of my source.

**Reply by MarioMey** [reply]
Yes, you are right: `aspect = 16.0/9.0;` does the trick.

It's a pity that we can't `print` any value to debug it...

**Comment by MarioMey** [comment]
Ok, fixed the aspect by @Limeth, I would want this shader to have some modifications. If anyone knows and wants to play a bit with it... 😁

- The ball/circle could be complete in screen, with no top/bottom crops.
- [DELETED, read EDIT]

With these modifications, we can move the source by the screen like a bubble.

[EDIT]: I change some code to make the ball opaque and the black ouside, transparent:

```glsl
		col *= vec4(
			pow(max(edge-len, 0.0), 0.2),
			pow(max(edge-dispersion-len, 0.0), 0.2),
			pow(max(edge-dispersion*2.0-len, 0.0), 0.2),
			pow(max(edge-len, 0.0), 0.05))*1.2;
```
